msctl

```diff
@@ -1274,6 +1277,123 @@ serverConsole() {
   done
 }
 
+# ---------------------------------------------------------------------------
+# Retrieve the timestamp.
+#
+# @return The current date and time.
+# ---------------------------------------------------------------------------
+timestamp() {
+  date +"%Y-%m-%d_%H-%M-%S"
+}
+
+# ---------------------------------------------------------------------------
+# Stop the server monitor.
+#
+# @param 1 The world server to stop.
+# ---------------------------------------------------------------------------
+stopServerMonitor() {
+  local WORLD_DIR MONITOR_LOG MONITOR_PID MONITOR_LOCK_FILE ACQUIRED_LOCK
+  WORLD_DIR="$WORLDS_LOCATION/$1"
+  MONITOR_LOG="$WORLD_DIR/logs/mscs.monitor.log"
+  MONITOR_PID="$WORLD_DIR/monitor.pid"
+  MONITOR_LOCK_FILE="$WORLD_DIR/monitor.lock"
+  # Check if server monitor instance currently running.
+  (
+  flock -n 9
+  ACQUIRED_LOCK=$?
+  if [ "$ACQUIRED_LOCK" -eq 1 ]; then # Server monitor is running.
+    printf "[$(timestamp)] [INFO]: Stop command received for server monitor. Attempting to kill server monitor...\n" >> "$MONITOR_LOG"
+    # Kill the server monitor.
+    kill -9 $(cat $MONITOR_PID)
+    # Verify it was actually killed.
+    if [ $? -eq 1 ]; then
+      printf "[$(timestamp)] [ERROR]: Unable to kill monitor process.\n" >> "$MONITOR_LOG"
+      exit 1
+    else 
+      printf "[$(timestamp)] [INFO]: Server monitor process killed sucessfully.\n" >> "$MONITOR_LOG"
+      # Remove the monitor PID file. 
+      rm -f "$WORLD_DIR/monitor.pid"
+    fi
+  fi
+  ) 9>"$MONITOR_LOCK_FILE"
+}
```

Comment:
This was kinda quick, but does this work?
```suggestion
  WORLD=$1
  MONITOR_LOCK_FILE=$WORLD_DIR/lock-monitor.pid
  # Check if server monitor instance currently running.
  if [ ! $(createLockFile $WORLD "monitor" ) ]  && [ -f "$MONITOR_LOCK_FILE" ]; then # Server monitor is running.
    printf "[$(timestamp)] [INFO]: Stop command received for server monitor. Attempting to kill server monitor...\n" >> "$MONITOR_LOG"
    # Kill the server monitor.
    kill -9 $(cat $MONITOR_LOCK_FILE)
    # Verify it was actually killed.
    if [ $? -eq 1 ]; then
      printf "[$(timestamp)] [ERROR]: Unable to kill monitor process.\n" >> "$MONITOR_LOG"
      exit 1
    else 
      printf "[$(timestamp)] [INFO]: Server monitor process killed successfully.\n" >> "$MONITOR_LOG"
      # Remove the monitor PID file. 
      rm -f "$MONITOR_LOCK_FILE"
    fi
  fi
}
```

Reply:
hm, what's `lock-monitor.pid`? Right now theres a file `flock` uses to maintain the lock called `monitor.lock`, this file is always there (that's just how flock works I believe). there's also a file called `monitor.pid` that stores the PID of the monitor.

Reply:
`lock-monitor.pid` is what `createLockFile` creates

Reply:
If you change `$(createLockFile $WORLD "monitor" )` to `$(createLockFile $WORLD "lock" )` the lock file would be `lock-lock.pid`

Reply:
so i think the issue with this is that the `createLockFile` function uses `$$` as the PID for the lock file, and `$$` expands to the PID of the current shell process. for the server monitor I'm generating a new process which has a _different PID_ than the current shell process. because of this, I believe that what happens is the lock files don't track the correct PID and the whole thing breaks.

the `createLockFile` function is fine for the use cases we have now--backups and overviewer mapping-- because it's checking to make sure that the _current shell PID_ isn't performing at multiple places at once, say, trying to perform a backup while a previous backup is still underway. it's not tracking multiple processes, it's just tracking one process--the current shell process. at least, thats my current understanding, lol.

`flock`, on the other hand, lets you track/lock multiple processes, and it manages all of the PID tracking for you. 

in conclusion, because  I'm creating a new process for the server monitor I don't think using `createLockFile` will work. and my testing validates this, I think--I did do it the way you described above initially when I was creating this PR and it didn't work, and I just tested it again with the changes you suggested and unfortunately it didn't work this time either.

this begs the question of why did I spin off a new process for the server monitor in the first place? well, I initially thought I wasn't going to do that, but then I realized that since it's a monitor i think it has to be a separate process, I dont think it can be the same process as the current shell.

let me know if you have any questions, this is just my understanding of the situation and I've been completely wrong before :blush:

thanks

Reply:
Makes sense, thanks for diving into it

Reply:
I think the only way to make `createLockFile` work is if you rewrite it.
I would hate to introduce another dependency for the script but it seems it could be fairly common.

> The flock command is part of the util-linux package and is available from Linux Kernel Archive (https://www.kernel.org/pub/linux/utils/util-linux/).

Either way, we should still add it to the `# Required Software` section and to the docs

Reply:
yea, i think flock is pretty common. it's also quite robust, so I'd feel comfortable using it